plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/portforwading/PortForwardingWizardModel.java

```diff
@@ -0,0 +1,190 @@
+					stream.println(NLS.bind("{0} {1} -> {2}", new Object [] {port.getName(), port.getLocalPort(), port.getRemotePort()}));
+				}
+				cap.forwardPorts(ports.toArray(new IPortForwardable.PortPair []{}));
+				stream.println("done.");
+				ConsoleUtils.displayConsoleView(console);
+				firePropertyChange(PROPERTY_PORT_FORWARDING, false, getPortForwarding());
+				return null;
+			}
+		}, null);
+	}
+	
+	private class ConsoleListener implements IConsoleListener{
+		private IPortForwardable cap;
+		
+		public void setCapability(IPortForwardable cap) {
+			this.cap = cap;
+		}
+		@Override
+		public void consolesRemoved(IConsole[] consoles) {
+			final String messageConsoleName = getMessageConsoleName();
+			for (IConsole console : consoles) {
+				if(console.getName().equals(messageConsoleName)) {
+					try {
+						cap.stop();
+						return;
+					}finally {
+						ConsoleUtils.deregisterConsoleListener(this);
+					}
+				}
+			}
+		}
+		
+		@Override
+		public void consolesAdded(IConsole[] consoles) {
+		}
+	}
+
+	private String getMessageConsoleName() {
+		return NLS.bind("Port forwarding to pod {0} ({1})", pod.getName(), pod.getNamespace());
+	}
+
+	public void stopPortForwarding() {
+		pod.accept(new CapabilityVisitor<IPortForwardable, Object>() {
+
+			@Override
+			public Object visit(IPortForwardable cap) {
+				if(!cap.isForwarding()) return null;
+				try {
+					final MessageConsole console = ConsoleUtils.findMessageConsole(getMessageConsoleName());
+					MessageConsoleStream stream = console.newMessageStream();
+					stream.println("Stopping port-forwarding...");
+					cap.stop();
+					for (IPortForwardable.PortPair port : ports) {
+						stream.println(NLS.bind("{0} {1} -> {2}", new Object [] {port.getName(), port.getLocalPort(), port.getRemotePort()}));
+					}
+					stream.println("done.");
+					ConsoleUtils.displayConsoleView(console);
+					firePropertyChange(PROPERTY_PORT_FORWARDING, true, getPortForwarding());
+					return null;
+				}finally {
+					ConsoleUtils.deregisterConsoleListener(consoleListener);					
+				}
+			}
+		}, null);
+			
+	}
+
+	public Boolean getUseFreePorts() {
+		return useFreePorts;
+	}
+
+	public void setUseFreePorts(Boolean useFreePorts) {
+		// do not change the current bindings if port forwarding is already started.
+		if (!getPortForwarding()) { 
+			updateLocalPortBindings(useFreePorts);
+		}
+		firePropertyChange(PROPERTY_USE_FREE_PORTS, this.useFreePorts, this.useFreePorts = useFreePorts);
+	}
+
+	private void updateLocalPortBindings(final boolean useFreePorts){
+		final List<String> bindings = new ArrayList<String>();
+		// update local bindings while avoiding duplicates
+		for (IPortForwardable.PortPair port : ports) {
+			if (useFreePorts) {
+				// find free port for every port
+				port.setLocalPort(SocketUtil.findFreePort());
+			} else {
+				// find duplicates and if match we find free port for those
+				// until stops.
+				port.setLocalPort(port.getRemotePort());
```

Comment:
@maxandersen wasnt there a special approach required for MacOS so that port forwarding works on MacOS? What was this? I vaguely remember that one could not bind to 127.0.0.1 but had to use the interface IP? Not sure if this was specific to ssh either?